s3ql: clean up dependencies, build for release, fix tests, fix `--systemd` option, fix `umount.s3ql`

justinlovinger opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

Move dependencies to proper inputs. Build for release instead of development. Fix and enable all tests. Use the correct `crypography` library instead of `pycrypto`, as per [installation instructions](http://www.rath.org/s3ql-docs/installation.html#installing-s3ql).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of (most) binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @rushmorem

JustinLovinger commented:
Merged commits.

P.S. Sorry for the notification spam. I accidentally branched from release-20.03 instead of master, pushing a number of extra commits. It's fixed now, but the bot already added reviewers.

jtojnar commented on pkgs/tools/backup/s3ql/default.nix:
```diff
@@ -11,25 +11,45 @@ python3Packages.buildPythonApplication rec {
-    # Removing integration tests
-    rm tests/t{4,5,6}_*
-    pytest tests
+    HOME=/build pytest tests
```
Not sure if /build is not an implementation detail, do not see it documented anywhere. Anyway, `$TMPDIR` might be a better choice.

JustinLovinger replied:
`HOME=/build` was a suggestion from sphalerite on NixOS IRC, to fix the `/homeless-shelter` issue: <https://logs.nix.samueldr.com/nixos/2019-08-26#;>. I'm not sure what exactly the suggestion was based on.

jtojnar replied:
It was based on the fact that the build is carried out in a writeable `/build` directory. But given that it is not documented anywhere and that it was changed in the past, I would consider it an implementation detail and use `HOME=$TMPDIR` instead.

lheckemann replied:
Oh no, why did I suggest that D: yes, I'm all for changing that.

JustinLovinger replied:
Done. Ironically `$TMPDIR` was `/build` when I tested it.

jtojnar commented on pkgs/tools/backup/s3ql/default.nix:
```diff
@@ -11,25 +11,45 @@ python3Packages.buildPythonApplication rec {
+    requests
+    systemd
+  ] ++ [
+    psmisc # `umount.s3ql` requires `fuser`
+    utillinuxMinimal.bin # `umount.s3ql` requires `umount`
```
This will not put those dependencies on `PATH`. You will need to patch the paths in https://github.com/s3ql/s3ql/blob/02b4b5b0775e3e9fb61581d4ab917aa351802b43/src/s3ql/umount.py.

JustinLovinger replied:
I'm not sure exactly how s3ql uses `fuser` and `umount`. Looking at my logs from the time I made this patch, `umount.s3ql` gave errors like `FileNotFoundError: [Errno 2] No such file or directory: 'fuser': 'fuser'`. Adding those packages to `buildInputs` did not fix the errors, but adding them to `propagatedBuildInputs` did. Someone who understands the role of `propagatedBuildInputs` better than myself may have better insight.

SuperSandro2000 replied:
Please use makeWrapper instead of abusing propagatedBuildInputs.

justinlovinger committed:
s3ql: clean up and fix

Clean up dependencies,
build for release,
fix tests,
fix `--systemd` option,
and fix `umount.s3ql`.

SuperSandro2000 commented on pkgs/tools/backup/s3ql/default.nix:
```diff
@@ -11,25 +11,45 @@ python3Packages.buildPythonApplication rec {
   };
 
-  checkInputs = [ which ] ++ (with python3Packages; [ cython pytest ]);
+  nativeBuildInputs = with python3Packages; [
+    cython
+  ];
+  buildInputs = [
+    sqlite
+  ];
+  checkInputs = with python3Packages; [
+    pytest
+  ] ++ [
+    which
+  ];
```
```suggestion
  nativeBuildInputs = [ python3Packages.cython ];

  buildInputs = [ sqlite ];

  checkInputs = [ python3Packages.pytest which ];
```

JustinLovinger commented:
I haven't used s3ql in a while. If anyone else wants to re-open this issue with the requested changes, feel free to use my code changes.
